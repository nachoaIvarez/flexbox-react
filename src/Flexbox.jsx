import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

// The number `0` with no unit is a valid CSS length, see:
//   https://www.w3.org/TR/CSS2/syndata.html#length-units
// Though it might be inappropriate/avoidable in certain contexts, if it is
// valid CSS, we should allow it.
const isTruthyOrZero = value => value || value === 0;

const Flexbox = styled(({
  alignContent,
  alignItems,
  alignSelf,
  children,
  display,
  element,
  flex,
  flexBasis,
  flexDirection,
  flexGrow,
  flexShrink,
  flexWrap,
  height,
  justifyContent,
  margin,
  marginBottom,
  marginLeft,
  marginRight,
  marginTop,
  maxHeight,
  maxWidth,
  minHeight,
  minWidth,
  order,
  padding,
  paddingBottom,
  paddingLeft,
  paddingRight,
  paddingTop,
  width,
  ...props
}) => React.createElement(element, props, children))`
  ${props => (props.alignContent ? `align-content: ${props.alignContent};` : '')}
  ${props => (props.alignSelf ? `align-self: ${props.alignSelf};` : '')}
  ${props => (props.alignItems ? `align-items: ${props.alignItems};` : '')}
  ${props => (props.display ? `display: ${props.display};` : '')}
  ${props => (isTruthyOrZero(props.flex) ? `flex: ${props.flex};` : '')}
  ${props => (isTruthyOrZero(props.flexBasis) ? `flex-basis: ${props.flexBasis};` : '')}
  ${props => (props.flexDirection ? `flex-direction: ${props.flexDirection};` : '')}
  ${props => (isTruthyOrZero(props.flexGrow) ? `flex-grow: ${props.flexGrow};` : '')}
  ${props => (isTruthyOrZero(props.flexShrink) ? `flex-shrink: ${props.flexShrink};` : '')}
  ${props => (props.flexWrap ? `flex-wrap: ${props.flexWrap};` : '')}
  ${props => (isTruthyOrZero(props.height) ? `height: ${props.height};` : '')}
  ${props => (props.justifyContent ? `justify-content: ${props.justifyContent};` : '')}
  ${props => (isTruthyOrZero(props.margin) ? `margin: ${props.margin};` : '')}
  ${props => (isTruthyOrZero(props.marginBottom) ? `margin-bottom: ${props.marginBottom};` : '')}
  ${props => (isTruthyOrZero(props.marginLeft) ? `margin-left: ${props.marginLeft};` : '')}
  ${props => (isTruthyOrZero(props.marginRight) ? `margin-right: ${props.marginRight};` : '')}
  ${props => (isTruthyOrZero(props.marginTop) ? `margin-top: ${props.marginTop};` : '')}
  ${props => (isTruthyOrZero(props.maxHeight) ? `max-height: ${props.maxHeight};` : '')}
  ${props => (isTruthyOrZero(props.maxWidth) ? `max-width: ${props.maxWidth};` : '')}
  ${props => (isTruthyOrZero(props.minHeight) ? `min-height: ${props.minHeight};` : '')}
  ${props => (isTruthyOrZero(props.minWidth) ? `min-width: ${props.minWidth};` : '')}
  ${props => (isTruthyOrZero(props.order) ? `order: ${props.order};` : '')}
  ${props => (isTruthyOrZero(props.padding) ? `padding: ${props.padding};` : '')}
  ${props => (isTruthyOrZero(props.paddingBottom) ? `padding-bottom: ${props.paddingBottom};` : '')}
  ${props => (isTruthyOrZero(props.paddingLeft) ? `padding-left: ${props.paddingLeft};` : '')}
  ${props => (isTruthyOrZero(props.paddingRight) ? `padding-right: ${props.paddingRight};` : '')}
  ${props => (isTruthyOrZero(props.paddingTop) ? `padding-top: ${props.paddingTop};` : '')}
  ${props => (isTruthyOrZero(props.width) ? `width: ${props.width};` : '')}
`;

Flexbox.propTypes = {
  alignContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
    'stretch',
  ]),
  alignItems: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  alignSelf: PropTypes.oneOf(['baseline', 'center', 'flex-end', 'flex-start', 'stretch']),
  children: PropTypes.node,
  display: PropTypes.oneOf(['flex', 'inline-flex']),
  element: PropTypes.oneOf([
    'article',
    'aside',
    'div',
    'figure',
    'footer',
    'header',
    'main',
    'nav',
    'section',
  ]),
  flex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  flexBasis: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  flexDirection: PropTypes.oneOf(['column-reverse', 'column', 'row-reverse', 'row']),
  flexGrow: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  flexShrink: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  flexWrap: PropTypes.oneOf(['nowrap', 'wrap-reverse', 'wrap']),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  inline: PropTypes.bool,
  justifyContent: PropTypes.oneOf([
    'center',
    'flex-end',
    'flex-start',
    'space-around',
    'space-between',
  ]),
  margin: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  marginBottom: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  marginLeft: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  marginRight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  marginTop: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  minHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  order: PropTypes.number,
  padding: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingBottom: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingLeft: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingRight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingTop: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  style: PropTypes.object,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Flexbox.defaultProps = {
  display: 'flex',
  element: 'div',
};

export default Flexbox;
