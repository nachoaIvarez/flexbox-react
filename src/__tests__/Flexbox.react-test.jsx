/* global document */

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Flexbox from '../Flexbox';

const TEST_CLASS = 'test-class';

const prettifyCSS = css =>
  css
    .replace(/\s+/g, '')
    .replace(/;/g, ';\n  ')
    .replace(/{/g, ' {\n  ')
    .replace(/ {2}}/g, '}\n')
    .replace(/:}/g, ': ');

const testComponent = component =>
  () => {
    const div = document.createElement('div');
    const mockComponent = renderer.create(component).toJSON();
    const classNames = mockComponent.props.className.split(' ');

    ReactDOM.render(component, div);
    const css = prettifyCSS(document.head.childNodes['0'].textContent);

    // Every class present in the rendered element must be either the TEST_CLASS (to test if
    // we can pass arbitrary classNames to the component) or generated classNames which must be
    // also present in the generated <style> rules by styled-components.
    expect(
      classNames.every(className => className === TEST_CLASS || css.includes(className)),
    ).toBeTruthy();
    expect(css).toMatchSnapshot();
    expect(mockComponent).toMatchSnapshot();
  };

test('Renders minimal <Flexbox /> component', testComponent(<Flexbox />));

test(
  'Renders custom element (header) <Flexbox /> component',
  testComponent(<Flexbox element="header" />),
);

test(
  'Renders <Flexbox /> component with several flexbox props',
  testComponent(
    <Flexbox
      alignItems="center"
      flexDirection="column"
      flexGrow={1}
      flexWrap="wrap"
      justifyContent="center"
    />,
  ),
);

test(
  'Renders <Flexbox /> component with special properties set to the number 0`',
  testComponent(
    <Flexbox
      flex={0}
      flexBasis={0}
      flexGrow={0}
      flexShrink={0}
      height={0}
      margin={0}
      marginBottom={0}
      marginLeft={0}
      marginRight={0}
      marginTop={0}
      maxWidth={0}
      maxHeight={0}
      minHeight={0}
      minWidth={0}
      order={0}
      padding={0}
      paddingBottom={0}
      paddingLeft={0}
      paddingRight={0}
      paddingTop={0}
      width={0}
    />,
  ),
);

test(
  'Renders <Flexbox /> component with layout helpers (height, width, margin, padding, etc)',
  testComponent(
    <Flexbox
      height="100%"
      margin="66px 33px"
      marginRight="25px"
      maxWidth="100vw"
      minHeight="100vh"
      padding="99px"
      paddingLeft="50px"
      width="100px"
    />,
  ),
);

test(
  'Renders <Flexbox /> with extra props, outside expected/defined ones (like event handlers or className)',
  testComponent(
    <Flexbox onClick={() => 'On click handler'} key="666" className="test-class" id="test-id" />,
  ),
);

test(
  'Renders <Flexbox /> with inline styles that can overwrite flexbox-react related props too',
  testComponent(
    <Flexbox
      style={{
        backgroundColor: 'red',
        display: 'block',
      }}
    />,
  ),
);
