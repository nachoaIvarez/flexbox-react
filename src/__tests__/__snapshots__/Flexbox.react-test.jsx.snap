// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Flexbox /> component with layout helpers (height, width, margin, padding, etc) 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
.DtNwl {
  -webkit-align-items:center;
  -webkit-box-align:center;
  -ms-flex-align:center;
  align-items:center;
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex-direction:column;
  -ms-flex-direction:column;
  flex-direction:column;
  -webkit-flex-grow:1;
  -ms-flex-grow:1;
  flex-grow:1;
  -webkit-flex-wrap:wrap;
  -ms-flex-wrap:wrap;
  flex-wrap:wrap;
  -webkit-box-pack:center;
  -webkit-justify-content:center;
  -ms-flex-pack:center;
  justify-content:center;
}
.jIWJwd {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex:0;
  -ms-flex:0;
  flex:0;
  -webkit-flex-basis:0;
  -ms-flex-basis:0;
  flex-basis:0;
  -webkit-flex-grow:0;
  -ms-flex-grow:0;
  flex-grow:0;
  -webkit-flex-shrink:0;
  -ms-flex-shrink:0;
  flex-shrink:0;
  height:0;
  margin:0;
  margin-bottom:0;
  margin-left:0;
  margin-right:0;
  margin-top:0;
  max-height:0;
  max-width:0;
  min-height:0;
  min-width:0;
  -webkit-order:0;
  -ms-flex-order:0;
  order:0;
  padding:0;
  padding-bottom:0;
  padding-left:0;
  padding-right:0;
  padding-top:0;
  width:0;
}
.eylexG {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  height:100%;
  margin:66px33px;
  margin-right:25px;
  max-width:100vw;
  min-height:100vh;
  padding:99px;
  padding-left:50px;
  width:100px;
}
"
`;

exports[`Renders <Flexbox /> component with layout helpers (height, width, margin, padding, etc) 2`] = `
<div
  className="sc-bdVaJa eylexG"
/>
`;

exports[`Renders <Flexbox /> component with several flexbox props 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
.DtNwl {
  -webkit-align-items:center;
  -webkit-box-align:center;
  -ms-flex-align:center;
  align-items:center;
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex-direction:column;
  -ms-flex-direction:column;
  flex-direction:column;
  -webkit-flex-grow:1;
  -ms-flex-grow:1;
  flex-grow:1;
  -webkit-flex-wrap:wrap;
  -ms-flex-wrap:wrap;
  flex-wrap:wrap;
  -webkit-box-pack:center;
  -webkit-justify-content:center;
  -ms-flex-pack:center;
  justify-content:center;
}
"
`;

exports[`Renders <Flexbox /> component with several flexbox props 2`] = `
<div
  className="sc-bdVaJa DtNwl"
/>
`;

exports[`Renders <Flexbox /> component with special properties set to the number 0\` 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
.DtNwl {
  -webkit-align-items:center;
  -webkit-box-align:center;
  -ms-flex-align:center;
  align-items:center;
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex-direction:column;
  -ms-flex-direction:column;
  flex-direction:column;
  -webkit-flex-grow:1;
  -ms-flex-grow:1;
  flex-grow:1;
  -webkit-flex-wrap:wrap;
  -ms-flex-wrap:wrap;
  flex-wrap:wrap;
  -webkit-box-pack:center;
  -webkit-justify-content:center;
  -ms-flex-pack:center;
  justify-content:center;
}
.jIWJwd {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex:0;
  -ms-flex:0;
  flex:0;
  -webkit-flex-basis:0;
  -ms-flex-basis:0;
  flex-basis:0;
  -webkit-flex-grow:0;
  -ms-flex-grow:0;
  flex-grow:0;
  -webkit-flex-shrink:0;
  -ms-flex-shrink:0;
  flex-shrink:0;
  height:0;
  margin:0;
  margin-bottom:0;
  margin-left:0;
  margin-right:0;
  margin-top:0;
  max-height:0;
  max-width:0;
  min-height:0;
  min-width:0;
  -webkit-order:0;
  -ms-flex-order:0;
  order:0;
  padding:0;
  padding-bottom:0;
  padding-left:0;
  padding-right:0;
  padding-top:0;
  width:0;
}
"
`;

exports[`Renders <Flexbox /> component with special properties set to the number 0\` 2`] = `
<div
  className="sc-bdVaJa jIWJwd"
/>
`;

exports[`Renders <Flexbox /> with extra props, outside expected/defined ones (like event handlers or className) 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
.DtNwl {
  -webkit-align-items:center;
  -webkit-box-align:center;
  -ms-flex-align:center;
  align-items:center;
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex-direction:column;
  -ms-flex-direction:column;
  flex-direction:column;
  -webkit-flex-grow:1;
  -ms-flex-grow:1;
  flex-grow:1;
  -webkit-flex-wrap:wrap;
  -ms-flex-wrap:wrap;
  flex-wrap:wrap;
  -webkit-box-pack:center;
  -webkit-justify-content:center;
  -ms-flex-pack:center;
  justify-content:center;
}
.jIWJwd {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex:0;
  -ms-flex:0;
  flex:0;
  -webkit-flex-basis:0;
  -ms-flex-basis:0;
  flex-basis:0;
  -webkit-flex-grow:0;
  -ms-flex-grow:0;
  flex-grow:0;
  -webkit-flex-shrink:0;
  -ms-flex-shrink:0;
  flex-shrink:0;
  height:0;
  margin:0;
  margin-bottom:0;
  margin-left:0;
  margin-right:0;
  margin-top:0;
  max-height:0;
  max-width:0;
  min-height:0;
  min-width:0;
  -webkit-order:0;
  -ms-flex-order:0;
  order:0;
  padding:0;
  padding-bottom:0;
  padding-left:0;
  padding-right:0;
  padding-top:0;
  width:0;
}
.eylexG {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  height:100%;
  margin:66px33px;
  margin-right:25px;
  max-width:100vw;
  min-height:100vh;
  padding:99px;
  padding-left:50px;
  width:100px;
}
"
`;

exports[`Renders <Flexbox /> with extra props, outside expected/defined ones (like event handlers or className) 2`] = `
<div
  className="test-class sc-bdVaJa cXgeQK"
  id="test-id"
  onClick={[Function]}
/>
`;

exports[`Renders <Flexbox /> with inline styles that can overwrite flexbox-react related props too 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
.DtNwl {
  -webkit-align-items:center;
  -webkit-box-align:center;
  -ms-flex-align:center;
  align-items:center;
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex-direction:column;
  -ms-flex-direction:column;
  flex-direction:column;
  -webkit-flex-grow:1;
  -ms-flex-grow:1;
  flex-grow:1;
  -webkit-flex-wrap:wrap;
  -ms-flex-wrap:wrap;
  flex-wrap:wrap;
  -webkit-box-pack:center;
  -webkit-justify-content:center;
  -ms-flex-pack:center;
  justify-content:center;
}
.jIWJwd {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  -webkit-flex:0;
  -ms-flex:0;
  flex:0;
  -webkit-flex-basis:0;
  -ms-flex-basis:0;
  flex-basis:0;
  -webkit-flex-grow:0;
  -ms-flex-grow:0;
  flex-grow:0;
  -webkit-flex-shrink:0;
  -ms-flex-shrink:0;
  flex-shrink:0;
  height:0;
  margin:0;
  margin-bottom:0;
  margin-left:0;
  margin-right:0;
  margin-top:0;
  max-height:0;
  max-width:0;
  min-height:0;
  min-width:0;
  -webkit-order:0;
  -ms-flex-order:0;
  order:0;
  padding:0;
  padding-bottom:0;
  padding-left:0;
  padding-right:0;
  padding-top:0;
  width:0;
}
.eylexG {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
  height:100%;
  margin:66px33px;
  margin-right:25px;
  max-width:100vw;
  min-height:100vh;
  padding:99px;
  padding-left:50px;
  width:100px;
}
"
`;

exports[`Renders <Flexbox /> with inline styles that can overwrite flexbox-react related props too 2`] = `
<div
  className="sc-bdVaJa cXgeQK"
  style={
    Object {
      "backgroundColor": "red",
      "display": "block",
    }
  }
/>
`;

exports[`Renders custom element (header) <Flexbox /> component 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
"
`;

exports[`Renders custom element (header) <Flexbox /> component 2`] = `
<header
  className="sc-bdVaJa cXgeQK"
/>
`;

exports[`Renders minimal <Flexbox /> component 1`] = `
"/*sc-component-id:sc-bdVaJa*/.sc-bdVaJa {
}
.cXgeQK {
  display:-webkit-box;
  display:-webkit-flex;
  display:-ms-flexbox;
  display:flex;
}
"
`;

exports[`Renders minimal <Flexbox /> component 2`] = `
<div
  className="sc-bdVaJa cXgeQK"
/>
`;
